File: 2680-76-1079-letter-tile-possibilities/2680-76-1079-letter-tile-possibilities.java

class Solution {
    boolean[] vis = new boolean[8];  
    
    private int solve(char[] tiles) {
        int ret = 1;
        for (int i = 0; i < tiles.length; i++) {
            if (vis[i] || (i > 0 && tiles[i] == tiles[i - 1] && !vis[i - 1])) continue;
            vis[i] = true;
            ret += solve(tiles);
            vis[i] = false;
        }
        return ret;
    }

    public int numTilePossibilities(String tiles) {
        char[] tileArray = tiles.toCharArray();
        Arrays.sort(tileArray); // Sort characters
        return solve(tileArray) - 1;
    }

}

File: 2680-76-1079-letter-tile-possibilities/2680-76-1079-letter-tile-possibilities.js
var numTilePossibilities = function(tiles) {
    tiles = tiles.split("").sort().join("");
    let vis = new Array(tiles.length).fill(false);
    
    function solve() {
        let ret = 1;
        for (let i = 0; i < tiles.length; i++) {
            if (vis[i] || (i > 0 && tiles[i] === tiles[i - 1] && !vis[i - 1])) continue;
            vis[i] = true;
            ret += solve();
            vis[i] = false;
        }
        return ret;
    }

    return solve() - 1;
};

